### DISCUSSION TITLE
Loops

### DISCUSSION THREAD

**Discussion by abirmingham** [markdoc/markdoc#308] [General]
Hello,

I'm currently evaluating Markdoc's fit for a project I am working on. It looks very promising. One piece of the documentation that jumped out at me is [Loops](https://markdoc.dev/docs/examples#loops), which indicates that loops require a custom Node or Tag `transform` function (or a custom react component). I am curious to understand more about why loop support is placed on the library user, versus in something like jinja where they are supported more out-of-the-box. It seems to me that this is a deliberate design decision, so I am sure that there is a reason for it. Can you talk a bit about this so that I can understand better?

Thanks!

**Comment by abirmingham** [comment]
Additionally - what is the correct way to write a generic foreach loop with Markdoc? I attempted to do so with a custom tag's `transform` function, but ran into the same issue as https://github.com/markdoc/markdoc/issues/154.

**Comment by rpaul-stripe** [comment]
Yes, you are correct that this is a deliberate design decision. The intent is to enforce separation of code and content by not supporting general-purpose template programming. This decision was informed by challenges we faced in our legacy documentation platform, where we had a lot of maintainability issues with loop-based procedural content generation in our content.

We benefited a lot from moving that kind of templating behavior out of the content and into the code, where test cases and type systems help to enforce correctness. It's quite straightforward to use a JavaScript loop inside of a tag `transform` function or a React component, fully encapsulating the content generation.

I'm not sure there is necessarily a good way to implement generic looping in Markdoc with the current architecture, due to the way that variables are resolved ahead of the transform step. If you look at the way that we implement [parameterization in partials](https://github.com/markdoc/markdoc/blob/main/src/tags/partial.ts#L44), that's basically what you want. You would have to defer the resolve step for everything nested inside of your loop tag and then recursively resolve it with your scoped variables for each iteration of the loop.

I think we feel pretty strongly about not including a loop construct in Markdoc out of the box, but if there's sufficient interest I'm open to coming up with some general pattern that custom tags can use to defer variable resolution for their children so that it's easy to create tags that implement loops or other behaviors that rely on having more control over variable resolution.

**Reply by abirmingham** [reply]
Got it. I can't weigh in on whether a feature that defers variable resolution has enough value to justify the complexity that it adds, but it certainly sounds interesting :)

At any rate, thanks for the thoughtful response @rpaul-stripe!

**Reply by m4rrc0** [reply]
Hey @rpaul-stripe . I am personally very interested to see an example of a recommended pattern when using only the tag `transform` function.

In the meantime, I implemented @johnagan's solution on #154 and it seems to work. Thanks a lot John!
